kartothek/io/dask/repartition.py

```diff
@@ -0,0 +1,93 @@
+import math
+from typing import Optional
+
+import dask
+import dask.dataframe as dd
+
+import kartothek.io.dask.dataframe as ktk_ddf
+from kartothek.core.factory import DatasetFactory
+from kartothek.serialization import PredicatesType
+
+
+def optimize(ddf: dd.DataFrame) -> dd.DataFrame:
+    """
+    Optimize a dask graph using the dask.Dataframe optimizer
+
+    Parameters
+    ----------
+    ddf
+        dask.Dataframe
+
+    Returns
+    --------
+    dask.DataFrame
+    """
+    return dask.optimize(ddf, optimizations=[dask.dataframe.optimize])[0]
+
+
+def calculate_repartition_ratio(
+    dataset_factory: DatasetFactory,
+    predicates: Optional[PredicatesType] = None,
+    *,
+    target_nrows_per_partition: int,
+    dataset_metadata_frac: float,
+) -> dd.core.Scalar:
+    """
+    Calculate a `repartition_ratio` according to the dataset input and `target_nrows_per_partition` (the desired
+    number of rows of each dataframe partition).
+
+
+    Parameters
+    ----------
+    dataset_factory
+        dataset factory
+    predicates
+        Optional , but predicates to be applied during calculatin repartition ratio
+    target_nrows_per_partition
+        target num of rows per partition
+    dataset_metadata_frac
+        used during collection of dataset metadata
+
+    Returns
+    -------
+    repartition_ratio
+        calculated repartition_ratio
+    """
+    dataset_metadata = ktk_ddf.collect_dataset_metadata(
+        factory=dataset_factory, predicates=predicates, frac=dataset_metadata_frac,
+    )
+    # Once the kartothek issue is resolved, one should use the rowgroup wise
+    # metadata. This would set up future improvements once kartothek is able to
+    # filter RG predicates while collecting statistics (assuming this will ever
+    # be possible this way)
+    # See https://github.com/JDASoftwareGroup/kartothek/pull/352
+    # and https://issues.apache.org/jira/browse/ARROW-10146
+    expected_nrows_per_partition = dataset_metadata[
+        ["partition_label", "number_rows_total"]
+    ].drop_duplicates()
```

Comment:
```suggestion
    expected_nrows_per_partition = dataset_metadata.groupby("partition_label")[
        "number_rows_per_row_group"
    ].sum()
```

Reply:
@usha-nemani-by let's do it future proof. While the groupby is technically more expensive, it shoundn't matter much. The drop dups is not cheap either. If we choose this implementation, future improvements in the statistics will automatically propagate to this function